src/ff1/alloc.rs

```diff
@@ -165,65 +164,189 @@ impl BinaryNumeralString {
-        self.0.len()
+        self.0.len() * 8
     }
 
     fn split(&self) -> (Self::Ops, Self::Ops) {
         let n = self.numeral_count();
         let u = n / 2;
         let v = n - u;
-        let mut front = self.0.clone();
-        let back = front.split_off(u);
-        (BinaryOps::new(front, u), BinaryOps::new(back, v))
-    }
-
-    fn concat(mut a: Self::Ops, mut b: Self::Ops) -> Self {
-        a.data.append(&mut b.data);
-        BinaryNumeralString(a.data)
+        let a_end = (u + 7) / 8;
+        let b_start = u / 8;
+
+        // FF1 processes the two halves of a numeral string as big-endian integers in the
+        // given radix, via the `NUM_radix()` operation. We are operating on binary data
+        // with a radix of 2, which means the "bit string" is interpreted as big endian.
+        //
+        // However, `BinaryNumeralString::from_bytes_le` uses little-endian bit order when
+        // parsing a byte encoding into a bit string (which indeed it should, otherwise
+        // the byte encoding would be mixed-endian which no one should have to suffer).
+        //
+        // The strategy taken in `FlexibleNumeralString` (which `BinaryNumeralString`
+        // previously also used) is to parse the little-endian byte string into (what is
+        // effectively) a `Vec<bool>`, and then read that as a big-endian bit pattern to
+        // compute the corresponding `BigUint` arithmetic value. For binary data we can do
+        // better, but we need to take care about how the data is parsed at each step.
+        //
+        // Say the input was 5 bytes (for the sake of illustration, so we can show both
+        // multiple bytes and how half-bytes / "nibbles" are handled). Let's draw out the
+        // bytes, annotated with the least and most significant bytes (LSB, MSB) and bits
+        // (lsb, msb), and the numeral string indices for each bit:
+        //
+        // LSB                                       MSB
+        //  | 0..7 | 8..15 | 16..23 | 24..31 | 32..39 |
+        // lsb    msb
+        //
+        // We need to split this into two pieces that have the same numeral string indices
+        // but *opposite* endianness interpretation of the numerals (lsn, msn):
+        //
+        //    msn      lsn
+        // a = |  0..19 |
+        // b = | 20..39 |
+        //
+        // We also want to store the bits so we can parse with `BigUint::from_bytes_le`
+        // (which avoids an unnecessary allocation per FF1 round). This means that we need
+        // the bits to be arranged within the sub-string bytes as follows:
+        //
+        // LSB                                 MSB
+        //  | 19..12 | 11...4 |  3...0 / [0; 4] |
+        //  | 39..32 | 31..24 | 23..20 / [0; 4] |
+        // lsb      msb
+        //
+        // If instead we were using a radix of 2^8 = 256, then we would be operating on a
+        // "byte string" and the bit ordering of each byte would not matter. Alas.
```

Comment:
I'm terribly sorry to have chosen 88 bits for the diversifier length. We should have chosen 96 bits. (I think the problem was that we chosen that length before we realized that we needed to use a PRP.)